content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt

```diff
@@ -0,0 +1,301 @@
+
+        // create the regex pattern to be used as first argument in `preg_replace_callback()`
+        $headlinesPattern = is_array($headlines) ? implode('|', $headlines) : $headlines;
+
+        // use `preg_replace_callback()` to replace matches with anchors
+        $text = preg_replace_callback('!<(' . $headlinesPattern . ')>(.*?)</\\1>!s', function ($match) {
+          
+            // create the id from the headline text
+            $id = Str::slug(Str::unhtml($match[2]));
+
+            // return the modified headline: 
+            // $match[1] contains the match for the first subpattern, i.e. `h2`, `h3` etc.
+            // $match[2] contains the match for the second subpattern, i.e. the actual headline text
+            return '<' . $match[1] . ' id="' . $id . '"><a href="#' . $id . '">' . $match[2] . '</a></' . $match[1] . '>';
+
+        }, $text);
+
+        return $text;
+      },
+    ]
+  ]
+]);
+```
+<info>
+Since we want to add a second hook of the same type in the next step, we use an array of functions after the `kirbytext:after` key.
+</info>
+
+The important part here is the `preg_replace_callback()` function. It needs three parameters…
+ 
+- the search pattern (`$headlinesPattern`)
+- the callback function
+- the input (`$text`)
+
+…and returns the modified string.
+
+To include more headlines levels, you can set the `k-cookbook.toc.headlines` option in the config:
+
+```php "/site/config/config.php"
+
+<?php
+
+return [
+  // …other settings
+  'k-cookbook.toc.headlines' => ['h2', 'h3', 'h4'],
+];
+```
+
+### Result
+If you now look at a page that contains headlines of the given level, you will see that the anchors were added.
+
+## Generate the ToC
+
+While the code we need to generate the ToCs from the headlines in a textarea field is the same, we can use different approaches to add them to the page.
+
+- Use a placeholder that users can manually put into a textarea field where they want the ToC to appear
+- Add a ToC in our template code using a field method
+
+We will look into both options in this recipe.
+
+<info>
+To keep it simple, we will include only one headline level in our ToC. We will cover multiple levels in a future recipe.
+</info>
+
+## Using a `toc` placeholder
+
+For the `toc` placeholder, we use another `kirbyText:after` hook, and a snippet that will contain the HTML for the ToC. We will reuse this snippet for the (link: #alternative-a-field-method-to-generate-the-toc text: alternative field method approach) below.
+
+```php "/site/plugins/toc/index.php"
+<?php
+
+use Kirby\Toolkit\Collection;
+use Kirby\Toolkit\Str;
+use Kirby\Toolkit\Obj;
+
+Kirby::plugin('k-cookbook/toc', [
+  'hooks' => [
+  
+    'kirbytext:after' => [
+
+      // (…) add the callback function from above here (without the wrapper)
+        
+      function($text) {
+
+        // the pattern allows passing an optional headline level `(toc: h3)`
+        $pattern  = '!\(toc:?\s?(h[1-6])?\)!';
```

Comment:
```suggestion
        $pattern  = '!\(toc(?::\s?(h[1-6]))?\)!';
```

(the colon is only optional when no parameter is passed – needs to be tested though if it still fully works)